src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsModel.cs

```diff
@@ -574,34 +600,50 @@ private void AddSchema(CsdlSchema schema, bool addAnnotations)
+        {
+            foreach (CsdlAnnotations schemaOutOfLineAnnotations in schema.OutOfLineAnnotations)
             {
-                this.SetNamespaceAlias(schema.Namespace, schema.Alias);
-            }
+                string target = this.ReplaceAlias(schemaOutOfLineAnnotations.Target);
 
-            if (addAnnotations)
-            {
-                foreach (CsdlAnnotations schemaOutOfLineAnnotations in schema.OutOfLineAnnotations)
+                List<CsdlSemanticsAnnotations> annotations;
+                if (!this.outOfLineAnnotations.TryGetValue(target, out annotations))
                 {
-                    string target = this.ReplaceAlias(schemaOutOfLineAnnotations.Target);
-
-                    List<CsdlSemanticsAnnotations> annotations;
-                    if (!this.outOfLineAnnotations.TryGetValue(target, out annotations))
-                    {
-                        annotations = new List<CsdlSemanticsAnnotations>();
-                        this.outOfLineAnnotations[target] = annotations;
-                    }
-
-                    annotations.Add(new CsdlSemanticsAnnotations(schemaWrapper, schemaOutOfLineAnnotations));
+                    annotations = new List<CsdlSemanticsAnnotations>();
+                    this.outOfLineAnnotations[target] = annotations;
                 }
+
+                CsdlAnnotations filteredAnnotations = FilterIncludedAnnotations(schemaOutOfLineAnnotations, target, includeAnnotations);
+                annotations.Add(new CsdlSemanticsAnnotations(schemaWrapper, filteredAnnotations));
             }
+        }
 
-            var edmVersion = this.GetEdmVersion();
-            if (edmVersion == null || edmVersion < schema.Version)
+        private CsdlAnnotations FilterIncludedAnnotations(CsdlAnnotations csdlAnnotations, string target, IEnumerable<IEdmIncludeAnnotations> includeAnnotations)
+        {
+            if (includeAnnotations == null || !includeAnnotations.Any())
             {
-                this.SetEdmVersion(schema.Version);
+                return csdlAnnotations;
             }
+
+            var filteredAnnotations = csdlAnnotations.Annotations.Where(annotation =>
+                ShouldIncludeAnnotation(annotation, target, includeAnnotations));
+            return new CsdlAnnotations(filteredAnnotations, csdlAnnotations.Target, csdlAnnotations.Qualifier);
+        }
+
+        private bool ShouldIncludeAnnotation(CsdlAnnotation annotation, string target, IEnumerable<IEdmIncludeAnnotations> includeAnnotations)
+        {
+            // include annotation if it matches one of the edmx:IncludeAnnotation references
+            // we check if the namespace matches, and (if applicable) also the target namespace and the qualifier
+            // see: http://docs.oasis-open.org/odata/odata-csdl-xml/v4.01/odata-csdl-xml-v4.01.html#sec_IncludedAnnotations
+
+            return includeAnnotations.Any(include =>

```

Comment:
>Any [](start = 38, length = 3)

LINQ is expensive -- AGS has gone through a LINQ hunt and removed all use of LINQ from their code.  We should avoid it where we can.

Instead, it would be more efficient to iterate through the collection and evaluate the options.